atom/browser/api/atom_api_net_log.cc

```diff
@@ -21,103 +22,145 @@
+  auto channel_string = std::string("Electron " ATOM_VERSION);
+  base::Value custom_constants = base::Value::FromUniquePtrValue(
+      net_log::ChromeNetLog::GetPlatformConstants(command_line_string,
+                                                  channel_string));
+
   auto* network_context =
       content::BrowserContext::GetDefaultStoragePartition(browser_context_)
           ->GetNetworkContext();
 
+  network_context->CreateNetLogExporter(mojo::MakeRequest(&net_log_exporter_));
+  net_log_exporter_.set_connection_error_handler(
+      base::BindOnce(&NetLog::OnConnectionError, base::Unretained(this)));
+
   // TODO(deepak1556): Provide more flexibility to this module
   // by allowing customizations on the capturing options.
-  net_log_writer_->StartNetLog(
-      log_path, net::NetLogCaptureMode::Default(),
-      net_log::NetExportFileWriter::kNoLimit /* file size limit */,
-      base::CommandLine::ForCurrentProcess()->GetCommandLineString(),
-      std::string(), network_context);
-}
+  auto capture_mode = network::mojom::NetLogCaptureMode::DEFAULT;
+  auto max_file_size = network::mojom::NetLogExporter::kUnlimitedFileSize;
 
-std::string NetLog::GetLoggingState() const {
-  if (!net_log_state_)
-    return std::string();
-  const base::Value* current_log_state =
-      net_log_state_->FindKeyOfType("state", base::Value::Type::STRING);
-  if (!current_log_state)
-    return std::string();
-  return current_log_state->GetString();
-}
+  base::PostTaskAndReplyWithResult(
+      file_task_runner_.get(), FROM_HERE,
+      base::BindOnce(OpenFileForWriting, log_path),
+      base::BindOnce(&NetLog::StartNetLogAfterCreateFile,
+                     weak_ptr_factory_.GetWeakPtr(), capture_mode,
+                     max_file_size, std::move(custom_constants),
+                     std::move(promise)));
 
-bool NetLog::IsCurrentlyLogging() const {
-  const std::string log_state = GetLoggingState();
-  return (log_state == "STARTING_LOG") || (log_state == "LOGGING");
+  return handle;
 }
 
-std::string NetLog::GetCurrentlyLoggingPath() const {
-  // Net log exporter has a default path which will be used
-  // when no log path is provided, but since we don't allow
-  // net log capture without user provided file path, this
-  // check is completely safe.
-  if (IsCurrentlyLogging()) {
-    const base::Value* current_log_path =
-        net_log_state_->FindKeyOfType("file", base::Value::Type::STRING);
-    if (current_log_path)
-      return current_log_path->GetString();
+void NetLog::StartNetLogAfterCreateFile(
+    network::mojom::NetLogCaptureMode capture_mode,
+    uint64_t max_file_size,
+    base::Value custom_constants,
+    util::Promise promise,
+    base::File output_file) {
+  if (!net_log_exporter_) {
+    // Theoretically the mojo pipe could have been closed by the time we get
+    // here via the connection error handler.
+    promise.RejectWithErrorMessage("Failed to create net log exporter");
```

Comment:
Wouldn't it be better to handle this in `NetLog::OnConnectionError` ? The main reason I want to do that is because `NetLog::StartNetLogAfterCreateFile` is a task invoked after a task in the file task runner is done with, but the mojo connection error handler is on the main task runner, and both these tasks aren't sequenced, so there can be a race.

Reply:
Yep, that's the reason to have this check here. Here's the sequence of operations this protects against:
![Untitled](https://user-images.githubusercontent.com/172800/58282848-b85b5500-7d5b-11e9-95cc-90db9e09410a.png)

`OnConnectionError` doesn't have access to the `Promise` so it can't reject the error.

Reply:
If `file created` task happens before `!!error!!` task, will it still be resolved to a failure ? Thats the case I was worried about.

> OnConnectionError doesn't have access to the Promise so it can't reject the error.

Yeah, I was imagining the `NetLog` class can manage the promise so that the rejection can happen in this method, but I am not sure if thats a good design.

Reply:
Hm.. if the file is created first, then we have a valid `net_log_exporter_` to send a `Start` message on. If the `error` task happens after that, the promise will never be resolved. I'm not sure what situation would cause that behaviour though... network service crash maybe? network context shutdown?

Reply:
> network service crash maybe? network context shutdown?

yup those are the possibilities I am imagining, just want to make sure we handle these scenarios gracefully.

Reply:
Perhaps this sequence of events could cause the promise to never be resolved:

1. Successfully create file
2. Send `Start` message to logger
3. Before the logger can report that it's started logging, there's a connection error on the mojo pipe (maybe because of a network service internal error of some kind, or possibly due to other things happening e.g. windows closing or shutdown in-progress)
4. the mojo pipe is closed and the promise is destroyed without being resolved

I guess in order to resolve this we need to stash the pending `startLogging` promise around somewhere and properly reject it in the connection error case.

Reply:
Yup, i was thinking to stash the promise in the `NetLog` class and use that to resolve. Based on our api design, seems like we have to deal with only one promise at a time so this might be possible ?